models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,469 @@
+
+  ArrayDatum tau_stc_list_ad( tau_stc_ );
+  def< ArrayDatum >( d, names::tau_stc, tau_stc_list_ad );
+
+  ArrayDatum q_stc_list_ad( q_stc_ );
+  def< ArrayDatum >( d, names::q_stc, q_stc_list_ad );
+}
+
+void
+nest::gif_psc_exp::Parameters_::set( const DictionaryDatum& d )
+{
+
+  updateValue< double >( d, names::I_e, I_e_ );
+  updateValue< double >( d, names::E_L, E_L_ );
+  updateValue< double >( d, names::g_L, g_L_ );
+  updateValue< double >( d, names::C_m, c_m_ );
+  updateValue< double >( d, names::V_reset, V_reset_ );
+  updateValue< double >( d, names::delta_u, delta_u_ );
+  updateValue< double >( d, names::v_t_star, v_t_star_ );
+  updateValue< double >( d, "lambda0", lambda0_ );
+  updateValue< double >( d, names::t_ref, t_ref_ );
+  updateValue< double >( d, names::tau_syn_ex, tau_ex_ );
+  updateValue< double >( d, names::tau_syn_in, tau_in_ );
+
+  updateValue< std::vector< double > >( d, names::tau_sfa, tau_sfa_ );
+  updateValue< std::vector< double > >( d, names::q_sfa, q_sfa_ );
+  updateValue< std::vector< double > >( d, names::tau_stc, tau_stc_ );
+  updateValue< std::vector< double > >( d, names::q_stc, q_stc_ );
+
+
+  if ( tau_sfa_.size() != q_sfa_.size() )
+    throw BadProperty( String::compose(
+      "'tau_sfa' and 'q_sfa' need to have the same dimensions.\nSize of "
+      "tau_sfa: %1\nSize of q_sfa: %2",
+      tau_sfa_.size(),
+      q_sfa_.size() ) );
+
+  if ( tau_stc_.size() != q_stc_.size() )
+    throw BadProperty( String::compose(
+      "'tau_stc' and 'q_stc' need to have the same dimensions.\nSize of "
+      "tau_stc: %1\nSize of q_stc: %2",
+      tau_stc_.size(),
+      q_stc_.size() ) );
+
+  if ( g_L_ <= 0 )
+    throw BadProperty( "Membrane conductance must be strictly positive." );
+
+  if ( delta_u_ <= 0 )
+    throw BadProperty( "delta_u must be strictly positive." );
+
+  if ( c_m_ <= 0 )
+    throw BadProperty( "Capacitance must be strictly positive." );
+
+  if ( t_ref_ < 0 )
+    throw BadProperty( "Refractory time must not be negative." );
+
+  for ( uint_t i = 0; i < tau_sfa_.size(); i++ )
+    if ( tau_sfa_[ i ] <= 0 )
+      throw BadProperty( "All time constants must be strictly positive." );
+
+  for ( uint_t i = 0; i < tau_stc_.size(); i++ )
+    if ( tau_stc_[ i ] <= 0 )
+      throw BadProperty( "All time constants must be strictly positive." );
+
+
+  if ( tau_ex_ <= 0 || tau_in_ <= 0 )
+    throw BadProperty( "Synapse time constants must be strictly positive." );
+}
+
+void
+nest::gif_psc_exp::State_::get( DictionaryDatum& d, const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y3_ );  // Membrane potential
+  def< double >( d, names::E_sfa, q_ ); // Adaptive threshold potential
+}
+
+void
+nest::gif_psc_exp::State_::set( const DictionaryDatum& d, const Parameters_& p )
+{
+  updateValue< double >( d, names::V_m, y3_ );
+  updateValue< double >( d, names::E_sfa, q_ );
+  initialized_ = false; // vectors of the state should be initialized with new parameter set.
```

Comment:
You are right. I removed `updateValue< double >( d, names::E_sfa, q_ )`. However, we need the variable `sfa_stc_initialized_`. Here is the story:
When a new set of parameter is given by the user, we need to reconstruct arrays of `P_sfa_`, `P_stc_`, `stc_elems_` and `sfa_elems` in the `calibrate` based on the new values of `sfa` and `stc` parameters. However, we cannot reconstruct them each time `calibrate` is called. For example, in the below situation
```
Simulate(10.0)
Simulate(10.0)
Simulate(10.0)
```
the arrays should be reconstructed only in the calibration of first `Simulate`. But in the case of 
```
Simulate(10.0)
SetStatus(...)
Simulate(10.0)
```
the arrays should be reconstructed in the calibration of both. Therefore, we need to inform the `calibrate` when it should do the reconstruction. The variable `sfa_stc_initialized_` keeps this information for us. When a new parameter set is given by the user, this variable is set to `false`. Note that we cannon change its value in `Parameters_::set` (non-static reference in a static method). Therefore we do that in `State_::set`.